### PR TITLE
Keep the context around for as long as you need it

### PR DESCRIPTION
**Pull request by gkokolatos**
A previous commit [1], rather embarrassingly, failed to notice the reset of used
memory contexts. Additionally it was missdiagnosed for PostgreSQL versions 10
and 11. A simple run with assertions enabled was enough to expose the real bug.

This commit ammends that error and in passing restructures a bit the code in
HealthCheckWorkerLauncherMain to be hopefully cleaner.

Now the launcherContext is still reset in every database scan, but objects that
need to persist, do reside in the parent context.

[1] commit 785fd533bb

Addresses issue #717

### CODE REVIEW

**Comment by JelteF** [comment]
This is definitly a lot better than before. I tried this locally and it allows me to run pg_auto_failover again when PG is compiled with assertions enabled. 

I'm not entirely sure this fixes everything though. Travis still randomly fails sometimes with "FATAL: the database is in recovery mode", which indicates that there has been a crash before:
```
test_multi_async.test_014_001_fail_node1 ... state of /tmp/multi_async/node2 is 'secondary', waiting for 'stop_replication' ...
ERROR
======================================================================
ERROR: test_multi_async.test_014_001_fail_node1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/python/3.7.6/lib/python3.7/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/citusdata/pg_auto_failover/tests/test_multi_async.py", line 241, in test_014_001_fail_node1
    assert node2.wait_until_state(target_state="stop_replication", timeout=120)
  File "/home/travis/build/citusdata/pg_auto_failover/tests/pgautofailover_utils.py", line 1153, in wait_until_state
    current_state, assigned_state = self.get_state()
  File "/home/travis/build/citusdata/pg_auto_failover/tests/pgautofailover_utils.py", line 1244, in get_state
    self.group,
  File "/home/travis/build/citusdata/pg_auto_failover/tests/pgautofailover_utils.py", line 368, in run_sql_query
    conn = psycopg2.connect(self.connection_string())
  File "/opt/python/3.7.6/lib/python3.7/site-packages/psycopg2/__init__.py", line 130, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
psycopg2.OperationalError: FATAL:  the database system is in recovery mode
````

**Comment by gkokolatos** [comment]
‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
On Tuesday, 8 June 2021 10:34, Jelte Fennema ***@***.***> wrote:

> This is definitly a lot better than before. I tried this locally and it allows me to run pg_auto_failover again when PG is compiled with assertions enabled.

Thanks for that!


>
> I'm not entirely sure this fixes everything though. Travis still randomly fails sometimes with "FATAL: the database is in recovery mode", which indicates that there has been a crash before:


I started to look at that myself. I have proven in an environment that is irrelevant to this commit.

Basically, if a primary node (node1 In this case) fails, then other nodes get promoted and change state. During this process,
a node that was secondary (node2) will be stopped, some configuration on it will change, and then get started again. This is
controlled by pg_autoctl. If the timing is such, while it is waiting for a state change, it will issue a connect command and it
might fail either with connection refused or the database is in recovery. Both of those are expected parts of the transition.

I am betting that some test instability exists around wait_*_state calls.

**Comment by gkokolatos** [comment]
‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
On Tuesday, June 8th, 2021 at 10:45, Dimitri Fontaine ***@***.***> wrote:

> > This is controlled by pg_autoctl. If the timing is such, while it is waiting for a state change, it will issue a connect command and it might fail either with connection refused or the database is in recovery. Both of those are expected parts of the transition. I am betting that some test instability exists around wait_*_state calls.
>
> That is surprising to me. In the test failures I had a look at, including the one here shared by Jelte, the query that fails to connect runs `self.monitor.run_sql_query(...)`: so the monitor Postgres instance has crashed. That's not expected, and this bug here looks like a smoking gun...

Fair enough. Let me do some more digging then.

**Comment by gkokolatos** [comment]
‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

On Tuesday, June 8th, 2021 at 10:48, ***@***.***> wrote:

> ‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
>
> On Tuesday, June 8th, 2021 at 10:45, Dimitri Fontaine ***@***.*** wrote:
>
> > > This is controlled by pg_autoctl. If the timing is such, while it is waiting for a state change, it will issue a connect command and it might fail either with connection refused or the database is in recovery. Both of those are expected parts of the transition. I am betting that some test instability exists around wait_*_state calls.
> >
> > That is surprising to me. In the test failures I had a look at, including the one here shared by Jelte, the query that fails to connect runs `self.monitor.run_sql_query(...)`: so the monitor Postgres instance has crashed. That's not expected, and this bug here looks like a smoking gun...
>
> Fair enough. Let me do some more digging then.

Debugging this was indeed fun. I mean it. Now I have to explain my findings which is not as straight forward.


I will try to do it here and I accept that I might fail to communicate well what I have found. Please correct me
where you think I am wrong.

The setup was as follows:

To start, I tried to run the test case in a loop until it fails. I am using the latest pg13 version.
Also I wanted to have the logs and data files available in the case that a failure happened. I modified
utils so that the teardown step will not remove them by removing the drop node:

--- a/tests/pgautofailover_utils.py
+++ b/tests/pgautofailover_utils.py
@@ -1109,10 +1109,9 @@ class DataNode(PGNode):
         try:
             destroy = PGAutoCtl(self)
             destroy.execute(
-                "pg_autoctl drop node --destroy",
+                "pg_autoctl drop node",
                 "drop",
                 "node",
-                "--destroy",
                 timeout=3,
             )
         except Exception as e:
@@ -1697,7 +1696,7 @@ class MonitorNode(PGNode):
         try:
             destroy = PGAutoCtl(self)
             destroy.execute(
-                "pg_autoctl destroy monitor", "drop", "monitor", "--destroy"
+                "pg_autoctl destroy monitor", "drop", "monitor"
             )
         except Exception as e:
             print(str(e))


Then I wrote a small bash function to remove the state/config files that pg_autoctl
is using along with the datafiles and named it remove_pgauf_state.

Finally I run the offending test in a loop like so:
$ for i in {1..100}; do remove_pgauf_state; TEST=test_multi_async make test || break; done;

True enough after a few successful runs, the test failed at the same point.


======================================================================
ERROR: test_multi_async.test_014_001_fail_node1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/georgios/pg_auto_failover/tests/test_multi_async.py", line 241, in test_014_001_fail_node1
    assert node2.wait_until_state(target_state="stop_replication", timeout=120)
  File "/home/georgios/pg_auto_failover/tests/pgautofailover_utils.py", line 1152, in wait_until_state
    current_state, assigned_state = self.get_state()
  File "/home/georgios/pg_auto_failover/tests/pgautofailover_utils.py", line 1236, in get_state
    results = self.monitor.run_sql_query(
  File "/home/georgios/pg_auto_failover/tests/pgautofailover_utils.py", line 368, in run_sql_query
    conn = psycopg2.connect(self.connection_string())
  File "/usr/lib/python3/dist-packages/psycopg2/__init__.py", line 126, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
psycopg2.OperationalError: FATAL:  the database system is in recovery mode



Great, now there is a reproduction, core files and log files.

Starting with the core, we see this:


#0  0x00007f3230e24cb8 in TupleToAutoFailoverNode (tupleDescriptor=0x55733a632e10, heapTuple=0x7f7f7f7f7f7f7f7f) at node_metadata.c:107
#1  0x00007f3230e2ddad in SetNodeHealthState (nodeId=4, nodeName=0x55733a5e1a20 "node_4", nodeHost=0x55733a5e1a40 "172.27.1.6", nodePort=5432, previousHealthState=1, healthState=0) at health_check_metadata.c:223
#2  0x00007f3230e2aab1 in ManageHealthCheck (healthCheck=0x55733a5e1c40, currentTime=...) at health_check_worker.c:850
#3  0x00007f3230e2a528 in DoHealthChecks (healthCheckList=0x55733a5e1c98) at health_check_worker.c:686
#4  0x00007f3230e2a25c in HealthCheckWorkerMain (arg=16385) at health_check_worker.c:577

and easy enough heapTuple is a nope: 0x7f7f7f7f7f7f7f7f

heapTuple comes from SPI_tuptable->vals[0]
which was set by SPI_execute


The code is found in SetNodeHealthState() and is the following:


    if (HaMonitorHasBeenLoaded())
    {
        initStringInfo(&query);
        appendStringInfo(&query,
                         "UPDATE " AUTO_FAILOVER_NODE_TABLE
                         "   SET health = %d, healthchecktime = now() "
                         " WHERE nodeid = %d "
                         "   AND nodehost = %s AND nodeport = %d "
                         " RETURNING node.*",
                         healthState,
                         nodeId,
                         quote_literal_cstr(nodeHost),
                         nodePort);

        pgstat_report_activity(STATE_RUNNING, query.data);

        spiStatus = SPI_execute(query.data, false, 0);
        Assert(spiStatus == SPI_OK_UPDATE_RETURNING);

        if (healthState != previousHealthState)
        {
            HeapTuple heapTuple = SPI_tuptable->vals[0];
            AutoFailoverNode *pgAutoFailoverNode =
                TupleToAutoFailoverNode(SPI_tuptable->tupdesc, heapTuple);


This codepath assumes that the UPDATE is successful by checking (asserting but is the same)
spiStatus. However, what it does not do, is to check the number of returning rows. In this
run SPI_processed is actually 0.

(gdb) p SPI_processed
$12 = 0

That actually means that the quals of the update did not match any of the rows.

True enough, the logs have captured that node_4 has actually being removed:

 [267769] STATEMENT:  SELECT pgautofailover.remove_node($1, $2)
 [267769] LOG:  Setting goal state of node 1 "node_1" (172.27.1.3:5432) to apply_settings after removing standby node 4 "node_4" (172.27.1.6:5432) from formation default.
 [267769] STATEMENT:  SELECT pgautofailover.remove_node($1, $2)
<snip>
 [265002] LOG:  background worker "pg_auto_failover monitor healthcheck worker pg_auto_failover" (PID 265013) was terminated by signal 11: Segmentation fault
 [265002] DETAIL:  Failed process was running: UPDATE pgautofailover.node   SET health = 0, healthchecktime = now()  WHERE nodeid = 4    AND nodehost = '172.27.1.6' AND nodeport = 5432  RETURNING node.*



Given the above, I understand that there are two processes involved and one of them removes an
entry from the table
    const char *deleteQuery =
        "DELETE FROM " AUTO_FAILOVER_NODE_TABLE
        " WHERE nodeid = $1";
and another one executes an UPDATE expecting definitely one result on the same table. There does
seem that there exists a window that the this will not hold true.


What have I missed?

**Comment by DimCitus** [comment]
> This codepath assumes that the UPDATE is successful by checking (asserting but is the same) spiStatus. However, what it does not do, is to check the number of returning rows. In this run SPI_processed is actually 0. (gdb) p SPI_processed $12 = 0 That actually means that the quals of the update did not match any of the rows. 
>True enough, the logs have captured that node_4 has actually being removed: [267769] STATEMENT: SELECT pgautofailover.remove_node($1, $2) [267769] 

> Given the above, I understand that there are two processes involved and one of them removes an entry from the table const char *deleteQuery = "DELETE FROM " AUTO_FAILOVER_NODE_TABLE " WHERE nodeid = $1"; and another one executes an UPDATE expecting definitely one result on the same table. There does seem that there exists a window that the this will not hold true. What have I missed?

Thanks for the debugging sessions @gkokolatos ; great detective work! As you're saying, with a repro that leaves hints behind that's a better use of our time, to thanks for taking to time to make that happen. I think your analysis is spot-on, and even if other problems might cause this kind of bug, this one has to be fixed. I'm going to cook a patch/PR for this one and then link it to this issue here. Thanks again!

**Comment by gkokolatos** [comment]
‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
On Tuesday, June 8th, 2021 at 17:01, Dimitri Fontaine ***@***.***> wrote:

> > This codepath assumes that the UPDATE is successful by checking (asserting but is the same) spiStatus. However, what it does not do, is to check the number of returning rows. In this run SPI_processed is actually 0. (gdb) p SPI_processed $12 = 0 That actually means that the quals of the update did not match any of the rows.
> >
> > True enough, the logs have captured that node_4 has actually being removed: [267769] STATEMENT: SELECT pgautofailover.remove_node($1, $2) [267769]
>
> > Given the above, I understand that there are two processes involved and one of them removes an entry from the table const char *deleteQuery = "DELETE FROM " AUTO_FAILOVER_NODE_TABLE " WHERE nodeid = $1"; and another one executes an UPDATE expecting definitely one result on the same table. There does seem that there exists a window that the this will not hold true. What have I missed?
>
> Thanks for the debugging sessions @gkokolatos ; great detective work! As you're saying, with a repro that leaves hints behind that's a better use of our time, to thanks for taking to time to make that happen. I think your analysis is spot-on, and even if other problems might cause this kind of bug, this one has to be fixed. I'm going to cook a patch/PR for this one and then link it to this issue here. Thanks again!

Thank you Dimitri. It is a pleasure to help.

I am looking forward to your PR. I have cooked something small to move things forward but better to wait for your solution first.

While at it, I have found the cause of the second failure in travis. I was able to reproduce at least once ever 10 runs.

The output in travis and indeed a failed run is:

4007test_multi_alternate_primary_failures.test_002_001_add_two_standbys ... state of /tmp/multi_alternate_primary_failures/node2 is 'secondary', done waiting
4008ERROR
4009
4010======================================================================
4011ERROR: test_multi_alternate_primary_failures.test_002_001_add_two_standbys
4012----------------------------------------------------------------------
4013Traceback (most recent call last):
4014  File "/opt/python/3.7.6/lib/python3.7/site-packages/nose/case.py", line 198, in runTest
4015    self.test(*self.arg)
4016  File "/home/travis/build/citusdata/pg_auto_failover/tests/test_multi_alternate_primary_failures.py", line 56, in test_002_001_add_two_standbys
4017    assert node2.has_needed_replication_slots()
4018  File "/home/travis/build/citusdata/pg_auto_failover/tests/pgautofailover_utils.py", line 1561, in has_needed_replication_slots
4019    if self.pgmajor() == 10:
4020  File "/home/travis/build/citusdata/pg_auto_failover/tests/pgautofailover_utils.py", line 607, in pgmajor
4021    self.pgversion()
4022  File "/home/travis/build/citusdata/pg_auto_failover/tests/pgautofailover_utils.py", line 597, in pgversion
4023    self.run_sql_query("show server_version_num")[0][0]
4024  File "/home/travis/build/citusdata/pg_auto_failover/tests/pgautofailover_utils.py", line 368, in run_sql_query
4025    conn = psycopg2.connect(self.connection_string())
4026  File "/opt/python/3.7.6/lib/python3.7/site-packages/psycopg2/__init__.py", line 130, in connect
4027    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
4028psycopg2.OperationalError: could not connect to server: Connection refused
4029	Is the server running on host "172.27.1.4" and accepting
4030	TCP/IP connections on port 5432?

Now what happens here is a small timing problem.

node2.has_needed_replication_slots() requires to check the version, which runs a query on that node
(call self.run_sql_query("show server_version_num")[0][0] the trace above). That node has been assigned
and reached state secondary yet the primary has not reached its primary state. So when there is a window
that the query will be executed while node2 is reloading config/state due to the change in node1.

By changing the order of the calls in that test, to call has_needed_replication_slots() after node1 has
finished the transition and the cluster is stable, I was able to execute the tests successfully more than
100 times in a row. Note that this pattern (calling has_needed_replication_slots after all nodes have
reached the desired state) is used throughout the rest of the suite.

In short, I think that this failure is not relevant to the current patch. I am more than happy to provide
with a patch to fix the testing if you agree.

**Comment by DimCitus** [comment]
```
4007test_multi_alternate_primary_failures.test_002_001_add_two_standbys ... state of /tmp/multi_alternate_primary_failures/node2 is 'secondary', done waiting 
4008ERROR 
4009 
4010====================================================================== 
4011ERROR: test_multi_alternate_primary_failures.test_002_001_add_two_standbys 
4012---------------------------------------------------------------------- 
4013Traceback (most recent call last): 
4014 File "/opt/python/3.7.6/lib/python3.7/site-packages/nose/case.py", line 198, in runTest 
4015 self.test(*self.arg) 
4016 File "/home/travis/build/citusdata/pg_auto_failover/tests/test_multi_alternate_primary_failures.py", line 56, in test_002_001_add_two_standbys 
4017 assert node2.has_needed_replication_slots() 
4018 File "/home/travis/build/citusdata/pg_auto_failover/tests/pgautofailover_utils.py", line 1561, in has_needed_replication_slots 
4019 if self.pgmajor() == 10: 
4020 File "/home/travis/build/citusdata/pg_auto_failover/tests/pgautofailover_utils.py", line 607, in pgmajor 
4021 self.pgversion() 
4022 File "/home/travis/build/citusdata/pg_auto_failover/tests/pgautofailover_utils.py", line 597, in pgversion 
4023 self.run_sql_query("show server_version_num")[0][0] 
4024 File "/home/travis/build/citusdata/pg_auto_failover/tests/pgautofailover_utils.py", line 368, in run_sql_query 
4025 conn = psycopg2.connect(self.connection_string()) 
4026 File "/opt/python/3.7.6/lib/python3.7/site-packages/psycopg2/__init__.py", line 130, in connect 
4027 conn = _connect(dsn, connection_factory=connection_factory, **kwasync) 
4028psycopg2.OperationalError: could not connect to server: Connection refused 
4029 Is the server running on host "172.27.1.4" and accepting 
4030 TCP/IP connections on port 5432? 
```

> Now what happens here is a small timing problem. node2.has_needed_replication_slots() requires to check the version, which runs a query on that node (call self.run_sql_query("show server_version_num")[0][0] the trace above). That node has been assigned and reached state secondary yet the primary has not reached its primary state. So when there is a window that the query will be executed while node2 is reloading config/state due to the change in node1. By changing the order of the calls in that test, to call has_needed_replication_slots() after node1 has finished the transition and the cluster is stable, I was able to execute the tests successfully more than 100 times in a row. Note that this pattern (calling has_needed_replication_slots after all nodes have reached the desired state) is used throughout the rest of the suite. In short, I think that this failure is not relevant to the current patch. I am more than happy to provide with a patch to fix the testing if you agree.

I see. For the sake of reducing Travis duties and wait time, I suppose it'd be quite nice to also have this test suite fix part of this very PR, if you'd agree to it?

**Review by DimCitus — APPROVED** [review]
LGTM

**Comment by DimCitus** [comment]
> I see. For the sake of reducing Travis duties and wait time, I suppose it'd be quite nice to also have this test suite fix part of this very PR, if you'd agree to it?

Well. Let me take that back. One PR per task is cleaner. I'll merge this one after an update, and please have a look at refactoring/fixing the test suite the way you said, in another PR?

**Comment by gkokolatos** [comment]
Sent from ProtonMail for iOS

On Tue, Jun 8, 2021 at 17:55, Dimitri Fontaine ***@***.***> wrote:

>> I see. For the sake of reducing Travis duties and wait time, I suppose it'd be quite nice to also have this test suite fix part of this very PR, if you'd agree to it?
>
> Well. Let me take that back. One PR per task is cleaner. I'll merge this one after an update, and please have a look at refactoring/fixing the test suite the way you said, in another PR?

It should absolutely be a different PR.
I didn’t phrase it well.
I will submit one soon(tm)
